# Environment Preparation

## Imports

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

#display
from tabulate               import tabulate
from IPython.core.display   import HTML
from IPython.display        import Image

# machine learning

# metrics

# handle pickle objects
import pickle

## Path Handlers

In [2]:
datapath = '../data/'
imagepath = '../images/'

## Helper Functions


In [3]:
def jupyter_settings():
    """ Otimiza configurações gerais, padronizanod tamanhos de plots, etc """
    %matplotlib inline
    #%pylab inline - not recommended
    plt.style.use( 'bmh' )
    plt.rcParams['figure.figsize'] = [14, 6]
    plt.rcParams['font.size'] = 24
    display( HTML( '<style>.container { width:100% !important; }</style>') )
    pd.options.display.max_columns = None
    pd.options.display.max_rows = None
    pd.set_option( 'display.expand_frame_repr', False )
    sns.set()
    pd.set_option('display.max_columns', 30)
    pd.set_option('display.max_rows', 30)
jupyter_settings()

# Data Collection


with sqlalchemy - works like a charm, for all dbs

with Psycopg2 - works as well, just for postgreSQL

## With sqlalchemy

In [11]:
from sqlalchemy import create_engine
from sqlalchemy_utils import database_exists, create_database
from postgre_credentials import *

In [13]:
# load from file postgre_credentials
url = 'postgresql://'+pg_user+':'+pg_passwd+'@'+pg_host+':'+pg_port+'/'+pg_db+''

if not database_exists(url):
    #create_database(url)
    print ("database does not exist!")
engine = create_engine(url, pool_size=50, echo=False)

In [14]:
#connected:
engine.url

postgresql://member:***@comunidade-ds-postgres.c50pcakiuwi3.us-east-1.rds.amazonaws.com:5432/comunidadedsdb

In [15]:
#select distinct schemas
query = "SELECT distinct(table_schema) from information_schema.tables"
df = pd.read_sql( query, engine )
df.head(500)

table_schema
0              public
1          pg_catalog
2  information_schema
3               pa004

In [16]:
#select tables from schema
query = "SELECT * FROM information_schema.tables WHERE table_schema = 'pa004'"
df = pd.read_sql( query, engine )
df.head(500)


table_catalog table_schema table_name  table_type self_referencing_column_name reference_generation user_defined_type_catalog user_defined_type_schema user_defined_type_name is_insertable_into is_typed commit_action
0  comunidadedsdb        pa004      users  BASE TABLE                         None                 None                      None                     None                   None                YES       NO          None
1  comunidadedsdb        pa004    vehicle  BASE TABLE                         None                 None                      None                     None                   None                YES       NO          None
2  comunidadedsdb        pa004  insurance  BASE TABLE                         None                 None                      None                     None                   None                YES       NO          None

In [17]:
query = """SELECT * FROM pa004.insurance """
df = pd.read_sql( query, engine )
df.head()

id  previously_insured  annual_premium  vintage  response
0   1                   0         40454.0      217         1
1   2                   0         33536.0      183         0
2   3                   0         38294.0       27         1
3   4                   1         28619.0      203         0
4   5                   1         27496.0       39         0

In [18]:
#build raw dataset:
query = """

SELECT *
FROM pa004.users u 
INNER JOIN pa004.vehicle v on u.id = v.id 
INNER JOIN pa004.insurance i ON u.id = i.id

"""
df = pd.read_sql( query, engine )
df.head()

id  gender  age  region_code  policy_sales_channel  id  driving_license vehicle_age vehicle_damage  id  previously_insured  annual_premium  vintage  response
0   7    Male   23         11.0                 152.0   7                1    < 1 Year            Yes   7                   0         23367.0      249         0
1  13  Female   41         15.0                  14.0  13                1    1-2 Year             No  13                   1         31409.0      221         0
2  18  Female   25         35.0                 152.0  18                1    < 1 Year             No  18                   1         46622.0      299         0
3  31  Female   26          8.0                 160.0  31                1    < 1 Year             No  31                   0          2630.0      136         0
4  39    Male   45          8.0                 124.0  39                1    1-2 Year            Yes  39                   0         42297.0      264         0

In [19]:
# remove ids from vehicle and insurance (same as from users)
df_raw = pd.read_sql( query, engine )
df_raw = df_raw.drop( df.columns[[5, 9]], axis=1 )
df_raw.head()

gender  age  region_code  policy_sales_channel  driving_license vehicle_age vehicle_damage  previously_insured  annual_premium  vintage  response
0    Male   23         11.0                 152.0                1    < 1 Year            Yes                   0         23367.0      249         0
1  Female   41         15.0                  14.0                1    1-2 Year             No                   1         31409.0      221         0
2  Female   25         35.0                 152.0                1    < 1 Year             No                   1         46622.0      299         0
3  Female   26          8.0                 160.0                1    < 1 Year             No                   0          2630.0      136         0
4    Male   45          8.0                 124.0                1    1-2 Year            Yes                   0         42297.0      264         0

In [21]:
df_raw.shape

(381109, 11)

In [22]:
df_raw.to_csv(datapath+'data_collection_done.csv', index=False)

## With psycopg2

In [23]:
import psycopg2 as pg
from postgre_credentials import *

In [26]:
# load from file postgre_credentials
def psycopg_connect():
    conn = pg.connect(    user = pg_user,
                          password = pg_passwd,
                          host = pg_host,
                          port = pg_port,
                          database = pg_db)
    return conn

In [28]:
#Check if conn obj created
conn = psycopg_connect()
#conn 

### Schema Query

In [ ]:
cursor = conn.cursor()

query_schema = """
    SELECT nspname
    FROM pg_catalog.pg_namespace
"""
cursor.execute( query_schema )

record = cursor.fetchall()
cursor.close()
conn.close()

In [ ]:
record

### Table Query

In [ ]:
conn = psycopg_connect()
cursor = conn.cursor()

query_tables = """
    SELECT tablename
    FROM pg_tables
    WHERE schemaname='pa004'
"""
cursor.execute( query_tables )

record = cursor.fetchall()
print (record)
cursor.close() #important
conn.close() #important

### Collect Data using psycopg2 (just sample)

In [ ]:
conn = psycopg_connect()
cursor = conn.cursor()

query_table_users = """
    SELECT *
    FROM pa004.users u
    WHERE u.age > 44
    limit 10
"""
cursor.execute( query_table_users )

record = cursor.fetchall()
print (record)
cursor.close() #important
conn.close() #important

In [ ]:
data = pd.DataFrame( record )
data.head()
#would need to rename colums each time!

### Collect Data using Pandas (use this)

In [ ]:
conn = psycopg_connect()

#sample query
query_table_users = """
    SELECT *
    FROM pa004.users u
    WHERE u.age > 44
    limit 10
"""

df = pd.read_sql( query_table_users, conn )
df.head()

In [ ]:
#build raw dataset:
query_table_users = """
    SELECT *
    FROM pa004.users u 
    LEFT JOIN pa004.vehicle v ON u.id = v.id 
    LEFT JOIN pa004.insurance i ON u.id = i.id 
"""

df = pd.read_sql( query_table_users, conn )
df.head()

In [ ]:
# remove ids from vehicle and insurance (same as from users)
df_raw = pd.read_sql( query_table_users, conn )
df_raw = df_raw.drop( df.columns[[5, 9]], axis=1 )
df_raw.head()

In [ ]:
df_raw.shape

In [ ]:
#same result with sqlarquemy, won't export again
#df_raw.to_csv(datapath+'data_collection_done.csv', index=False)

# Data Description

In [ ]:
df1 = pd.read_csv(datapath+'data_collection_pandas_done.csv')
df1.sample(5)

In [ ]:
df1.info()

## Rename Columns

In [ ]:
df1.columns

In [ ]:
df1 = df1.rename(columns={'Unnamed: 0': 'id',
                        'policy_sales_channel': 'id_contact_code',
                        'previously_insured': 'vehicle_previously_insured',
                        'annual_premium': 'health_annual_premium',
                        'vintage': 'days_health_insured',
                        'response': 'vehicle_insurance_response'
                        })
df1

## Column Meanings

In [ ]:
#Double check understanding of each column

tab = [['Column', 'Meaning'],
    #####################################################  
        ['id', 'unique customer identifier.'],
        ['gender', 'client gender: Male / Female.'],
        ['age', 'customer age.'],
        ['region_code', 'customer region code.'],
        ['id_contact_code', 'anonymous code for customer contact channel.'],
        ['driving_license', '0 = client is not allowed to drive, 1 = has permission'],
        ['vehicle_age', 'vehicle age: < 1 Year, 1-2 Year, > 2 Years'],
        ['vehicle_damage', 'No = customer has never had their vehicle damaged in the past, Yes = has had it.'],
        ['vehicle_previously_insured', '0 = customer does not have vehicle insurance, 1= already has vehicle insurance.'],
        ['health_annual_premium', 'annual amount paid by the customer to the company for health insurance. Currency: Rs(Pakistani rupee, R$1.00 = ± Rs0.03).'],
        ['days_health_insured', 'number of days since the customer joined the company by purchasing health insurance. The policy is annual.'],
        ['vehicle_insurance_response', '0 = customer is not interested, 1 = customer is interested.']
      ]
print(tabulate(tab, headers='firstrow', stralign='left', tablefmt='pipe'))
#tablefmt='pipe', tablefmt='grid'

## Data Dimension

In [ ]:
print(f'Number of rows: {df1.shape[0]} ')
print(f'Number of columns: {df1.shape[1]} ')

## Data Types

In [ ]:
df1.dtypes

In [ ]:
df1

In [ ]:
#convert df['region_code'] to int:
#df['region_code'].value_counts()
df1['region_code'] = df1['region_code'].astype(int)
df1['region_code'].head()

In [ ]:
#convert df['id_contact_code'] to int:
#df['id_contact_code'].value_counts().head(50)
df1['id_contact_code'] = df1['id_contact_code'].astype(int)
df1['id_contact_code'].head()

In [ ]:
df.dtypes

## Check NA

In [ ]:
df.isna().sum()
#nothing to handle

## Fill NA

In [ ]:
#Not necessary
pass

## Descriptive Statistical

In [ ]:
df1.head()

In [ ]:
#separate cathegorical and numerical attributes
num_attributes = df1[['age','health_annual_premium','days_health_insured']]
cat_attributes = df1[['gender','region_code','id_contact_code','driving_license', 'vehicle_age', 'vehicle_damage', 'vehicle_previously_insured', 'vehicle_insurance_response']]

In [ ]:
num_attributes.sample(5)

In [ ]:
cat_attributes.sample(5)

### Numerical Attributes

In [ ]:
#Central tendency metrics: mean and median
ct1 = pd.DataFrame( num_attributes.apply( np.mean ) ).T
ct2 = pd.DataFrame( num_attributes.apply( np.median ) ).T

#Dispersion metrics: standard deviation, min, max, range, skew, kurtosis
d1 = pd.DataFrame ( num_attributes.apply ( np.std ) ).T
d2 = pd.DataFrame ( num_attributes.apply ( min ) ).T
d3 = pd.DataFrame ( num_attributes.apply ( max ) ).T
d4 = pd.DataFrame ( num_attributes.apply ( lambda x: x.max() - x.min() ) ).T
d5 = pd.DataFrame ( num_attributes.apply ( lambda x: x.skew() ) ).T
d6 = pd.DataFrame ( num_attributes.apply ( lambda x: x.kurtosis() ) ).T

#concatenate
m = pd.concat( [ d2, d3, d4, ct1, ct2, d1, d5, d6 ] ).T.reset_index()
m.columns = ['attributes', 'min', 'max', 'range', 'mean', 'median', 'std', 'skew', 'kurtosis' ]

In [ ]:
plt.subplot(2,2,1)
sns.histplot(df1['age'], binwidth=1, kde=True);
plt.subplot(2,2,2)
sns.histplot(df1['health_annual_premium'], kde=True);
plt.subplot(2,2,3)
sns.histplot(df1['days_health_insured'], binwidth=1, kde=True);

In [ ]:
m

Hilights:

age: mean = 36y, 1 std = 15.5  -->  68% of ages are between 20,5 and 51,5 y

health_annual_premium: bery big range of 537535, high kurtosis, so many extreme values.

days_health_insured: The policy is for 12 months, thats why the max number of days a customer is health insured is 299. 

### Cathegorical Attributes

In [ ]:
#variations of each column
print( cat_attributes.apply( lambda x: x.unique() ) ,'\n')
print( cat_attributes.apply( lambda x: x.unique().shape[0] ) )

In [ ]:
plt.subplots_adjust(top=1.5)

plt.subplot (3,2,1)
sns.countplot (x = cat_attributes['gender']);

plt.subplot (3,2,2)
sns.countplot (x = cat_attributes['driving_license']);
plt.xticks([0,1],['No','Yes']);

plt.subplot (3,2,3)
sns.countplot (x = cat_attributes['vehicle_age']);

plt.subplot (3,2,4)
sns.countplot (x = cat_attributes['vehicle_damage']);

plt.subplot (3,2,5)
sns.countplot (x = cat_attributes['vehicle_previously_insured']);
plt.xticks([0,1],['No','Yes']);

plt.subplot (3,2,6)
sns.countplot (x = cat_attributes['vehicle_insurance_response']);
plt.xticks([0,1],['No','Yes']);

In [ ]:
round(cat_attributes['vehicle_insurance_response'].value_counts() / cat_attributes['vehicle_insurance_response'].shape[0] * 100 ,2)

Highlights
Just 12% of customers answered to be interested in a vehicle insurance.

In [ ]:
#save dataset 
df1.to_csv(datapath+'data_description_done.csv')

# Feature Engeneering

In [ ]:
df2 = pd.read_csv(datapath+'data_collection_pandas_done.csv')
df2.head()

In [ ]:
df1

## Hypothesis mindmap

In [ ]:
#hypothesis mindmap
Image( imagepath+'mindmap_hypothesis.png' )

## Hypothesis creation

In [ ]:
#Created hypothesis with available dataset data
tab = [['ID', 'Hypothesis'],
################################################################################
        ['H1', 'Customer with AGE HIGHER should be MORE interested.'],
        ['H2', 'Cussomer with GENDER FEMALE should be MORE interested.'],
        ['H3', 'Customers of CERTAINS REGION should be MORE interested.'],
        ['H4', 'Customers with DRIVERS LICENCE should be MORE interested.'],
        ['H5', 'Customers with MOST RECENT VEHICLE AGE should be MORE interested.'],
        ['H6', 'Customers who HAD VEHICLE DAMAGE should be MORE interested.'],
        ['H7', 'Customers who HAD VEHICLE PREVIOUSLY INSURED should be MORE interested.'],
        ['H8', 'Customers with HIGHER ANNUAL HEALTH EXPANSES should be MORE interested'],
        ['H9', 'Customers whth MORE DAYS OF HEALTH INSURANCE should be MORE interested']]
################################################################################
print(tabulate(tab, headers='firstrow', stralign='left', tablefmt='pipe')) #tablefmt='pipe', tablefmt='grid'

## Hypothesis priorization

In [ ]:
#All hypothesis can be analysed on first CRISP cycle
pass

## Feature engeneering


In [ ]:
df.sample(5)

In [ ]:
dict_vehicle_damage = { '> 2 Years':'over_2_years', '1-2 Year':'between_1_2_years', '< 1 Year': 'below_1_year' }
dict_vehicle_damage
